### ISSUE TITLE
Error registering

### ISSUE THREAD

**Issue by d5** [aws/amazon-ecs-agent#135]
Whenever I start many ecs-agent installed EC2 instances (like more than 50) at the same time, some of them fail to register with the following errors:

```
# docker ps -a
CONTAINER ID        IMAGE                            COMMAND             CREATED             STATUS                     PORTS               NAMES
0e1015495546        amazon/amazon-ecs-agent:latest   "/agent"            3 minutes ago       Exited (1) 3 minutes ago                       ecs-agent           
# docker logs 0e1015495546
2015-07-11T02:29:34Z [INFO] Starting Agent: Amazon ECS Agent - v1.3.0 (097e4af)
2015-07-11T02:29:34Z [INFO] Loading configuration
2015-07-11T02:29:34Z [INFO] Checkpointing is enabled. Attempting to load state
2015-07-11T02:29:34Z [INFO] Loading state! module="statemanager"
2015-07-11T02:29:34Z [INFO] Registering Instance with ECS
2015-07-11T02:29:39Z [ERROR] Could not register module="api client" err="Post https://ecs.us-west-2.amazonaws.com/: net/http: request canceled while waiting for connection"
2015-07-11T02:29:39Z [ERROR] Error registering: Post https://ecs.us-west-2.amazonaws.com/: net/http: request canceled while waiting for connection
```

Is this a known issue? Any chances these failed agents retry registering again?

**Comment by euank** [comment]
Apologies for the slow response.

This issue has a few possible solutions. First of all, #136 updated the SDK we use which includes providing retries for API requests.
This might catch some network issues like this.

The second solution is to handle agent failures by restarting it. The ECS Optimized AMI's init script will relaunch the Agent when it exits (excluding exit codes 0 and 5).
This behavior can be added when running it with systemd by adding the following to your unit file:
```
Restart=on-failure
RestartPreventExitStatus=5
```

Depending on how you're launching the agent, there are other ways you can handle that as well.

I'll update this issue when we release a new Agent and I'd be happy to hear then if you still encounter these issues then.

Best,
Euan

**Comment by samuelkarp** [comment]
We've released [v1.3.1](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.3.1), which we believe has addressed this problem.  Please feel free to reopen or file a new issue if the problem persists.

Thanks!
Sam
